### PR TITLE
Exporter for Databricks SQL objects

### PR DESCRIPTION
**Pull request by alexott**

### CODE REVIEW

File: sql/resource_sql_endpoint.go [hunk 1]
```diff
@@ -99,9 +99,13 @@ type SQLEndpointsAPI struct {
 }
 
 // List all SQL endpoints
-func (a SQLEndpointsAPI) List() (l EndpointList, err error) {
-	err = a.client.Get(a.context, "/sql/endpoints", nil, &l)
-	return
+func (a SQLEndpointsAPI) List() ([]SQLEndpoint, error) {
+	var lst endpointList
```
Comment (nfx) [inline sql/resource_sql_endpoint.go:103]: Newer API wrappers have listing methods have struct returned, not arrays. Let's keep consistency.

File: sql/resource_sql_endpoint.go [hunk 2]
```diff
@@ -121,6 +125,9 @@ func (a SQLEndpointsAPI) Stop(endpointID string) error {
 // Get ...
 func (a SQLEndpointsAPI) Get(endpointID string) (se SQLEndpoint, err error) {
 	err = a.client.Get(a.context, fmt.Sprintf("/sql/endpoints/%s", endpointID), nil, &se)
+	if se.Tags != nil && len(se.Tags.CustomTags) == 0 {
```
Comment (nfx) [inline sql/resource_sql_endpoint.go:128]: Add a comment why this change is there

File: sql/resource_sql_endpoint_test.go [hunk 3]
```diff
@@ -280,7 +280,7 @@ func TestSQLEnpointAPI(t *testing.T) {
 		a := NewSQLEndpointsAPI(ctx, client)
 		list, err := a.List()
 		require.NoError(t, err)
-		assert.Len(t, list.Endpoints, 1)
+		assert.Len(t, list, 1)
```
Comment (nfx) [inline sql/resource_sql_endpoint_test.go:283]: Revert this one

**Review by nfx — CHANGES_REQUESTED** [review]
Some consistency comments, otherwise great feature to add.

I'll defer review of widget resource to @pietern

File: exporter/importables.go [hunk 4]
```diff
@@ -35,6 +37,49 @@ var (
 	globalWorkspaceConfName = "global_workspace_conf"
 )
 
+type sqlaListResponse struct {
```
Comment (pietern) [inline exporter/importables.go:40]: `s/sqla/dbsql/`

File: exporter/importables.go [hunk 5]
```diff
@@ -35,6 +37,49 @@ var (
...
+	Results    []map[string]interface{} `json:"results"`
+	Page       int64                    `json:"page"`
+	TotalCount int64                    `json:"count"`
+	PageSize   int64                    `json:"page_size"`
+}
+
+// Generic function to list objects related to the SQL Analytics
```
Comment (pietern) [inline exporter/importables.go:47]: DBSQL

File: exporter/importables.go [hunk 6]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+			return nil
+		},
+		Import: func(ic *importContext, r *resource) error {
+			if ic.meAdmin {
+				ic.Emit(&resource{
+					Resource: "databricks_permissions",
+					ID:       fmt.Sprintf("/sql/queries/%s", r.ID),
+					Name:     "sql_query_" + r.Data.Get("name").(string),
+				})
+			}
```
Comment (pietern) [inline exporter/importables.go:989]: Nice!

File: sql/resource_widget.go [hunk 7]
```diff
@@ -268,6 +268,16 @@ func ResourceWidget() *schema.Resource {
...
+			p, err := common.SchemaPath(m, "position", "pos_x")
+			if err == nil {
+				p.Optional = true
+				p.Required = false
+			}
+			p, err = common.SchemaPath(m, "position", "pos_y")
+			if err == nil {
+				p.Optional = true
+				p.Required = false
+			}
```
Comment (pietern) [inline sql/resource_widget.go:280]: Both of these are required in the payload, AFAIK. Why default to 0?

**Review by pietern — COMMENTED** [review]
Awesome to see you're working on this, @alexott !

File: exporter/importables.go (hunk 5 above)
Comment (alexott) [reply exporter/importables.go:47]: will fix - it was started when it was SQL Analytics :-)

File: sql/resource_widget.go (hunk 7 above)
Comment (pietern) [reply sql/resource_widget.go:280]: Maybe because `0` is the same as unset for non-pointer values?

Comment (alexott) [reply sql/resource_widget.go:280]: yes, but they should be ignored only if they are declared as `omitempty`, and this is not a case

File: exporter/importables.go [hunk 8]
```diff
@@ -35,6 +37,49 @@ var (
...
+	TotalCount int64                    `json:"count"`
+	PageSize   int64                    `json:"page_size"`
+}
+
+// Generic function to list objects related to the DBSQL
+func dbsqlListObjects(ic *importContext, path string) ([]map[string]interface{}, error) {
+	var listResponse dbsqlListResponse
+	err := ic.Client.Get(ic.Context, path, nil, &listResponse)
+	if err != nil {
+		return nil, err
```
Comment (nfx) [inline exporter/importables.go:52]: cover by a simple test? :)

File: exporter/importables.go [hunk 9]
```diff
@@ -35,6 +37,49 @@ var (
...
+	}
+	startPos := 0
+	curPos := len(listResponse.Results)
+	copy(events[startPos:curPos], listResponse.Results)
+	for curPos < totalCount {
+		err := ic.Client.Get(ic.Context, path,
+			map[string]interface{}{"page_size": listResponse.PageSize, "page": listResponse.Page + 1},
+			&listResponse)
+		if err != nil {
+			return nil, err
```
Comment (nfx) [inline exporter/importables.go:68]: cover by a simple test? :)

File: exporter/importables.go [hunk 10]
```diff
@@ -35,6 +37,49 @@ var (
...
+		}
+		startPos = curPos
+		curLen := len(listResponse.Results)
+		restItems := totalCount - startPos
+		if restItems < curLen {
+			curLen = restItems
```
Comment (nfx) [inline exporter/importables.go:74]: cover by a simple test? :)

File: exporter/importables.go [hunk 11]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
 	},
+	"databricks_sql_query": {
+		Service: "sql",
+		Name: func(d *schema.ResourceData) string {
+			return d.Get("name").(string) + "_" + d.Id()
+		},
+		List: func(ic *importContext) error {
+			qs, err := dbsqlListObjects(ic, "/preview/sql/queries")
+			if err != nil {
+				return nil
```
Comment (nfx) [inline exporter/importables.go:957]: cover by a simple test? :)

File: exporter/importables.go [hunk 12]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+					Resource: "databricks_sql_query",
+					ID:       q["id"].(string),
+				})
+				data_sources[q["data_source_id"].(string)] = struct{}{}
+				log.Printf("[INFO] Imported %d of %d SQL queries", i+1, len(qs))
+			}
+			for data_source := range data_sources {
+				endpoint_id, err := ic.getSqlEndpoint(data_source)
+				if err == nil {
+					ic.Emit(&resource{
```
Comment (nfx) [inline exporter/importables.go:971]: maybe for consistency/readability reasons, change it to:

```go
if err != nil {
  log...
  continue
}
```

File: exporter/importables.go [hunk 13]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+		Depends: []reference{
+			{Path: "data_source_id", Resource: "databricks_sql_endpoint", Match: "data_source_id"},
+		},
+	},
+	"databricks_sql_endpoint": {
+		Service: "sql",
+		Name: func(d *schema.ResourceData) string {
+			name := d.Get("name").(string)
+			if name == "" {
+				name = d.Id()
```
Comment (nfx) [inline exporter/importables.go:1001]: cover by a simple test? :)

File: exporter/importables.go [hunk 14]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+			}
+			return name
+		},
+		List: func(ic *importContext) error {
+			endpointsList, err := sql.NewSQLEndpointsAPI(ic.Context, ic.Client).List()
+			if err != nil {
+				return err
```
Comment (nfx) [inline exporter/importables.go:1008]: cover by a simple test? :)

File: exporter/importables.go [hunk 15]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+	"databricks_sql_visualization": {
+		Service: "sql",
+		Name: func(d *schema.ResourceData) string {
+			name := d.Get("name").(string) + "_" + d.Id()
+			return name
+		},
+		List: func(ic *importContext) error {
+			allVis, err := ic.getSqlVisualizations()
+			if err != nil {
+				return err
```
Comment (nfx) [inline exporter/importables.go:1039]: cover by a simple test? :)

File: exporter/importables.go [hunk 16]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+	},
+	"databricks_sql_dashboard": {
+		Service: "sql",
+		Name: func(d *schema.ResourceData) string {
+			return d.Get("name").(string) + "_" + d.Id()
+		},
+		List: func(ic *importContext) error {
+			qs, err := dbsqlListObjects(ic, "/preview/sql/dashboards")
+			if err != nil {
+				return nil
```
Comment (nfx) [inline exporter/importables.go:1064]: cover by a simple test? :)

File: exporter/importables.go [hunk 17]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+		Service: "sql",
+		Name: func(d *schema.ResourceData) string {
+			return d.Id()
+		},
+		List: func(ic *importContext) error {
+			visualizations := map[string]struct{}{}
+			dashboards := map[string]struct{}{}
+			dashboardList, err := dbsqlListObjects(ic, "/preview/sql/dashboards")
+			if err != nil {
+				return err
```
Comment (nfx) [inline exporter/importables.go:1097]: cover by a simple test? :)

File: exporter/importables.go [hunk 18]
```diff
@@ -901,4 +946,212 @@ var resourcesMap map[string]importable = map[string]importable{
...
+					cnt++
+				}
+			}
+
+			allVis, err := ic.getSqlVisualizations()
+			if err == nil {
+				for vis := range visualizations {
+					visID, ok := allVis[vis]
+					if ok {
+						ic.Emit(&resource{
```
Comment (nfx) [inline exporter/importables.go:1131]: eek.... such much nesting, can't read. can you reduce nesting to minimum? :)

File: sql/resource_sql_endpoint_test.go [hunk 19]
```diff
@@ -231,8 +232,8 @@ func TestSQLEnpointAPI(t *testing.T) {
 		{
 			Method:   "GET",
 			Resource: "/api/2.0/sql/endpoints",
-			Response: EndpointList{
```
Comment (nfx) [inline sql/resource_sql_endpoint_test.go:234]: consistency. revert this one :)

**Review by nfx — CHANGES_REQUESTED** [review]
few more comments

File: exporter/importables.go (hunk 8 above)
Comment (alexott) [reply exporter/importables.go:52]: Too many test cases really…

**Comment by alexott** [comment]
Don’t have capacity to increase code coverage, maybe sometime in April…

**Review by nfx — APPROVED** [review]
